Sprint 12

UlyanaHanush opened the pull request:
1.  Реализована постраничная загрузка данных для ленты фотографий
2. Реализована функциональность лайков
3. При тапе на картинку осуществляется переход на экран, где можно посмотреть полноразмерную версию фотографии
4. Добавлена возможность выхода из аккаунта
5. Добавлено логирование ошибок

UlyanaHanush committed:
feat(ImagesListService): logic for fetch photos

UlyanaHanush committed:
fix(ImagesListService): test completed successfully

UlyanaHanush committed:
feat(ImagesListViewController): displaying photos according to content ImagesListService,
 feat(SingleImageViewController): displaying photo according to content ImagesListService

UlyanaHanush committed:
feat(ImagesListService): func changeLike, makeChangeLikeRequest
feat(ImagesListViewController): func imageListCellDidTapLike

UlyanaHanush committed:
feat(ProfileLogoutService): added and implied in other service

UlyanaHanush committed:
fix(ImagesListCell): func prepareForReuse

UlyanaHanush committed:
feat(ImagesListCell): func setIsLiked
feat(ProfileViewController): didTapButton - alert
feat(ImagesListViewController): dateLabel.text from network

UlyanaHanush committed:
feat(StoryBoard): deleted

Helvetii951 commented on ImageFeed/ImageListView/ImagesListCell.swift:
```diff
@@ -67,6 +75,28 @@ final class ImagesListCell: UITableViewCell {
+    @IBAction func didTapLikeButton(_ sender: Any) {
+        delegate?.imageListCellDidTapLike(self)
+    }
+    
```
`didTapLikeButton ` можно сделать приватной.

Helvetii951 commented on ImageFeed/ImageListView/ImagesListViewController.swift:
```diff
@@ -108,32 +138,94 @@ extension ImagesListViewController: UITableViewDelegate {
+        
+        let url = URL(string: photos[indexPath.row].thumbImageURL)
+        let processor = RoundCornerImageProcessor(cornerRadius: 16)
+            
```
`RoundCornerImageProcessor` из `Kingfisher` скругляет именно ту картинку, которую скачал. Т.е., если он скачал большую картинку(например, 2000х2000), то он скруглит у нее только 16 поинтов, что вряд ли будет заметно. Всегда учитывай это при использовании `RoundCornerImageProcessor`.

Helvetii951 commented on ImageFeed/ImageListView/ImagesListViewController.swift:
```diff
@@ -108,32 +138,94 @@ extension ImagesListViewController: UITableViewDelegate {
+            case .failure(let error):
+                print("[ImagesListViewController]:[imageListCellDidTapLike] \(error.localizedDescription)")
+                UIBlockingProgressHUD.dismiss()
+                self.showAlert()
```
`UIBlockingProgressHUD.dismiss()` можно поднять на уровень `switch`. Так будет чуть меньше кода.

Helvetii951 commented on ImageFeed/ImageListView/Service/ImagesListService.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    private func makeChangeLikeRequest(photoId: String,  isLike: Bool) -> URLRequest? {
+        guard let url = URL(string: "https://api.unsplash.com/photos/\(photoId)/like"),
+              let token = oAuth2TokenStorage.token
```
Лучше брать базовый URL из констант.

Helvetii951 commented on ImageFeed/ProfileView/Service/ProfileLogoutService.swift:
```diff
@@ -0,0 +1,64 @@
+        imagesListService.cleanImagesList()
+    }
+    
+    private func switchToSplashViewController() {
+        // Получаем экземпляр `window` приложения
+        guard let window = UIApplication.shared.windows.first else {
+            fatalError("Invalid Configuration")
+        }
+        
+        let splashViewController = SplashViewController()
+        window.rootViewController = splashViewController
+    }
```
Обязанность по навигации в текущей архитектуре(MVC) лучше отдать контроллеру. Пусть сервис занимается только бизнес-логикой. Сообщаем через замыкание/делегата/уведомление, что разлогинились, а контроллер уже решает что делать.

Helvetii951 commented on ImageFeed/AppDelegate.swift:
```diff
@@ -7,12 +7,16 @@
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
+        ProgressHUD.animationType = .activityIndicator
+        ProgressHUD.colorHUD = .ypBlack
+        ProgressHUD.colorAnimation = .lightGray
```
Лучше выносить подобные настройки в отдельную функцию. Иначе [application(_:didFinishLaunchingWithOptions:)](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622921-application) может стать громоздкой.

А еще лучше было бы перенести в статическую функцию в обертку `UIBlockingProgressHUD`.

Helvetii951 commented on ImageFeed/ImageListView/ImagesListViewController.swift:
```diff
@@ -47,9 +50,20 @@ final class ImagesListViewController: UIViewController {
+    
+    // MARK: - Public Methods
+    
+    func showSingleImage(indexPath: IndexPath) {
+        let singleImageViewController = SingleImageViewController()
+        singleImageViewController.imageObject = imagesListService.photos[indexPath.row]
+        singleImageViewController.modalPresentationStyle = .fullScreen
+        present(singleImageViewController, animated: true, completion: nil)
     }
```
Тоже можно сделать `private`. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

UlyanaHanush committed:
fix(struct Photo): remove DateFormatter
fix(ImagesListService): func fetchPhotosNextPage() and func changeLike: add DateFormatter
fix(ImagesListViewController): func configCell: dateLabel.text

Helvetii951 commented on ImageFeed/ImageListView/ImagesListViewController.swift:
```diff
@@ -108,32 +136,95 @@ extension ImagesListViewController: UITableViewDelegate {
+        let photo = photos[indexPath.row]
         
-        let likeImage = isLiked ? UIImage(named: "like_button_on") : UIImage(named: "like_button_off")
+        guard let photoCreatedAt = photo.createdAt else {
+            return cell.dateLabel.text = nil
+        }
```
_Повторное ревью_

Здесь лучше воспользоваться if-let-else. Этим guard'ом ты можешь отрезать настройку других свойств ячейки.

Helvetii951 commented on ImageFeed/ImageListView/Service/ImagesListService.swift:
```diff
@@ -0,0 +1,148 @@
+                        id: result.id,
+                        size: CGSize(width: result.width, height: result.height),
+                        createdAt: {
+                            guard let dateString = result.createdAt else {
+                                return nil
+                            }
```
_Повторное ревью_

Здесь guard может полностью прервать выполнение функции. Рекомендую применить if-let.

UlyanaHanush committed:
fix(ImagesListService): func fetchPhotosNextPage: changed guard to let
fix(ImagesListViewController): func configCell: changed guard to let
